src/compiler/resolutionCache.ts

```diff
@@ -0,0 +1,316 @@
+
+    /*@internal*/
+    export function resolveWithGlobalCache(primaryResult: ResolvedModuleWithFailedLookupLocations, moduleName: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost, globalCache: string | undefined, projectName: string): ResolvedModuleWithFailedLookupLocations | undefined {
+        if (!isExternalModuleNameRelative(moduleName) && !(primaryResult.resolvedModule && extensionIsTypeScript(primaryResult.resolvedModule.extension)) && globalCache !== undefined) {
+            // otherwise try to load typings from @types
+
+            // create different collection of failed lookup locations for second pass
+            // if it will fail and we've already found something during the first pass - we don't want to pollute its results
+            const { resolvedModule, failedLookupLocations } = loadModuleFromGlobalCache(moduleName, projectName, compilerOptions, host, globalCache);
+            if (resolvedModule) {
+                return { resolvedModule, failedLookupLocations: primaryResult.failedLookupLocations.concat(failedLookupLocations) };
+            }
+        }
+    }
+
+    /*@internal*/
+    export function createResolutionCache(
+        toPath: (fileName: string) => Path,
+        getCompilerOptions: () => CompilerOptions,
+        watchForFailedLookupLocation: (failedLookupLocation: string, containingFile: string, name: string) => FileWatcher,
+        log: (s: string) => void,
+        resolveWithGlobalCache?: ResolverWithGlobalCache): ResolutionCache {
+
+        let host: ModuleResolutionHost;
+        let filesWithChangedSetOfUnresolvedImports: Path[];
+        let filesWithInvalidatedResolutions: Map<true>;
+
+        const resolvedModuleNames = createMap<Map<ResolvedModuleWithFailedLookupLocations>>();
+        const resolvedTypeReferenceDirectives = createMap<Map<ResolvedTypeReferenceDirectiveWithFailedLookupLocations>>();
+
+        type FailedLookupLocationsWatcher = { fileWatcher: FileWatcher; refCount: number };
+        const failedLookupLocationsWatches = createMap<FailedLookupLocationsWatcher>();
+
+        return {
+            setModuleResolutionHost,
+            startRecordingFilesWithChangedResolutions,
+            finishRecordingFilesWithChangedResolutions,
+            resolveModuleNames,
+            resolveTypeReferenceDirectives,
+            invalidateResolutionOfDeletedFile,
+            invalidateResolutionOfChangedFailedLookupLocation,
+            createHasInvalidatedResolution,
+            clear
+        };
+
+        function setModuleResolutionHost(updatedHost: ModuleResolutionHost) {
+            host = updatedHost;
+        }
+
+        function clear() {
+            failedLookupLocationsWatches.forEach((failedLookupLocationWatcher, failedLookupLocationPath: Path) => {

```

Comment:
Don't annotate the types of parameters, that acts like a downcast due to bivariance.